video/out/vo_drm.c

```diff
@@ -0,0 +1,523 @@
+            // check whether this CRTC works with the encoder
+            if (!(enc->possible_crtcs & (1 << j)))
+                continue;
+
+            drmModeFreeEncoder(enc);
+            dev->crtc = res->crtcs[j];
+            return 0;
+        }
+
+        drmModeFreeEncoder(enc);
+    }
+
+    MP_ERR(vo, "Connector %u has no suitable CRTC\n", conn->connector_id);
+    return -ENOENT;
+}
+
+static void modeset_destroy_fb(int fd, struct modeset_buf *buf)
+{
+    struct drm_mode_destroy_dumb dreq;
+
+    munmap(buf->map, buf->size);
+
+    drmModeRmFB(fd, buf->fb);
+
+    memset(&dreq, 0, sizeof(dreq));
+    dreq.handle = buf->handle;
+    drmIoctl(fd, DRM_IOCTL_MODE_DESTROY_DUMB, &dreq);
+}
+
+static bool is_connector_valid(struct vo *vo, int conn_id,
+                               drmModeConnector *conn, bool silent)
+{
+    if (!conn) {
+        if (!silent) {
+            char *errstr = mp_strerror(errno);
+            MP_ERR(vo, "Cannot get connector %d: %s\n", conn_id, errstr);
+        }
+        return false;
+    }
+
+    if (conn->connection != DRM_MODE_CONNECTED) {
+        if (!silent) {
+            MP_ERR(vo, "Connector %d is disconnected\n", conn_id);
+        }
+        return false;
+    }
+
+    if (conn->count_modes == 0) {
+        if (!silent) {
+            MP_ERR(vo, "Connector %d has no valid modes\n", conn_id);
+        }
+        return false;
+    }
+
+    return true;
+}
+
+static int modeset_prepare_dev(struct vo *vo, int fd, int conn_id,
+                               struct modeset_dev **out)
+{
+    struct modeset_dev *dev = NULL;
+    drmModeConnector *conn = NULL;
+    drmModeRes *res = NULL;
+    int ret = 0;
+
+    *out = NULL;
+
+    res = drmModeGetResources(fd);
+    if (!res) {
+        char *errstr = mp_strerror(errno);
+        MP_ERR(vo, "Cannot retrieve DRM resources: %s\n", errstr);
+        ret = -errno;
+        goto end;
+    }
+
+    if (conn_id == -1) {
+        // get the first connected connector
+        for (int i = 0; i < res->count_connectors; i ++) {
+            conn = drmModeGetConnector(fd, res->connectors[i]);
+            if (is_connector_valid(vo, i, conn, true)) {
+                conn_id = i;
+                break;
+            }
+            if (conn) {
+                drmModeFreeConnector(conn);
+                conn = NULL;
+            }
+        }
+        if (conn_id == -1) {
+            MP_ERR(vo, "No connected connectors found\n");
+            ret = -ENODEV;
+            goto end;
+        }
+    }
+
+    if (conn_id < 0 || conn_id >= res->count_connectors) {
+        MP_ERR(vo, "Bad connector ID. Max valid connector ID = %u\n",
+               res->count_connectors);
+        ret = -ENODEV;
+        goto end;
+    }
+
+    conn = drmModeGetConnector(fd, res->connectors[conn_id]);
+    if (!is_connector_valid(vo, conn_id, conn, true)) {
+        ret = -ENODEV;
+        goto end;
+    }
+
+    dev = talloc_size(vo->priv, sizeof(*dev));
```

Comment:
Also, for defensive programming, ``talloc_zero(vo->priv, struct modeset_dev)`` would be better.